cores/arduino/stm32/LowPower.c

```diff
@@ -0,0 +1,289 @@
+  * @attention
+  *
+  * <h2><center>&copy; COPYRIGHT(c) 2017 STMicroelectronics</center></h2>
+  *
+  * Redistribution and use in source and binary forms, with or without modification,
+  * are permitted provided that the following conditions are met:
+  *   1. Redistributions of source code must retain the above copyright notice,
+  *      this list of conditions and the following disclaimer.
+  *   2. Redistributions in binary form must reproduce the above copyright notice,
+  *      this list of conditions and the following disclaimer in the documentation
+  *      and/or other materials provided with the distribution.
+  *   3. Neither the name of STMicroelectronics nor the names of its contributors
+  *      may be used to endorse or promote products derived from this software
+  *      without specific prior written permission.
+  *
+  * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+  * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+  * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+  * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+  * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+  * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+  * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+  * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+  * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+  * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+  *
+  ******************************************************************************
+  */
+
+#include "Arduino.h"
+#include "LowPower.h"
+
+#ifdef HAL_PWR_MODULE_ENABLED
+
+#ifdef __cplusplus
+ extern "C" {
+#endif
+
+static UART_HandleTypeDef* WakeUpUart = NULL;
+static void (*WakeUpUartCb)( void ) = NULL;
+
+/**
+  * @brief  Initialize low power mode
+  * @param  None
+  * @retval None
+  */
+void LowPower_init(){
+  /* Enable Power Clock */
+  __HAL_RCC_PWR_CLK_ENABLE();
+
+  /* Ensure that HSI is wake-up system clock */
+  __HAL_RCC_WAKEUPSTOP_CLK_CONFIG(RCC_STOP_WAKEUPCLOCK_HSI);
+
+  /* Check if the system was resumed from StandBy mode */
+  if (__HAL_PWR_GET_FLAG(PWR_FLAG_SB) != RESET)
+  {
+    /* Clear Standby flag */
+    __HAL_PWR_CLEAR_FLAG(PWR_FLAG_SB);
+  }
+
+  /* Clear all related wakeup flags */
+  __HAL_PWR_CLEAR_FLAG(PWR_FLAG_WU);
+}
+
+/**
+  * @brief  Configure a pin as wakeup source if compatible.
+  * @param  pin: pin to configure
+  * @param  mode: pin mode (edge or state). The configuration have to be compatible.
+  * @retval None
+  */
+void LowPower_EnableWakeUpPin(uint32_t pin, uint32_t mode) {
+#ifndef STM32L4xx
+  UNUSED(mode);
+#endif
+  PinName p = digitalPinToPinName(pin);
+  uint32_t wupPin = pinmap_find_function(p, PinMap_WKUP);
```

Comment:
PinMap_WKUP is the config? Maybe it should be at sketch level ? And/Or adding a method to set wakeup pins?

Reply:
PinMap_WKUP is the list of wake up pins available on the board.
I don't think it should be at sketch level.
I suggest to move PinMap_WKUP in variant.cpp. What do you think about it?

Reply:
Wake up pins touch the standby mode.

If you confirm us that pins are the same across a MCU family, we will able to look for another logic and remove the array from variant.cpp.

Reply:
Please add it in the comments or above table declaration that PinMap_WKUP only applies to standby/shutdown.  I am not sure whether pins are always the same accross MCUs from the same family ...